### DISCUSSION TITLE
How to share variables between HTTP requests in FastAPI?

### DISCUSSION THREAD

**Discussion by NathaliaBarreiros** [fastapi/fastapi#4617] [Questions] [answered]
I wonder how can I share the value of variables between HTTP request in FastAPI. For instance, I have a post request in which I get some audio files and then I convert their info into a pandas dataframe, I would like to send that dataframe in a GET request, but I can't access to the dataframe on the GET request scope.

```
@app.post(
    path="/upload-audios/",
    status_code=status.HTTP_200_OK
)
async def upload_audios(audios: list[UploadFile] = File(...)):
    filenames = [audio.filename for audio in audios]
    audio_data = [audio.file for audio in audios]
    new_data = []
    final_data = []
    header = ["name", "file"]
    for i in range(len(audios)):
        new_data = [filenames[i], audio_data[i]]
        final_data.append(new_data)
    new_df = pd.DataFrame(final_data, columns=header)
    return f"You have uploaded {len(audios)} audios which names are: {filenames}"

@app.get("/get-dataframe/")
async def get_dataframe():
    pass
```

**Comment by jgould22** [comment]
You would use a database for something like this.

**Comment by BilalAlpaslan** [comment]
you can declare variables in upper scope

**Comment by sm-Fifteen** [comment, marked as answer]
That depends on what you're trying to do.

* If you're only looking at returning some metadata on the uploaded file, you could probably without the following GET.
* If you were looking to process the audio data and returning the processed files, you can't return more than one file with an HTTP request, so returning all of them as the response of the post wouldn't be an option.
* If you wanted the uploaded files to be permanently uploaded and hosted by your application, you would need some kind of database to manage that file data. Regarding the specific use case of chaining a POST with a GET, you're looking for something closer to an HTTP `201 Created` response.

```
@app.post(
    path="/upload-audio/",
    status_code=status.HTTP_201_CREATED
)
async def upload_audio(response: Response, audio: UploadFile = File(...)):
    # Put the file data in the database
    # Get the inserted id of the file

    response.headers['Location'] = f'/get-dataframe/{new_file_id}'
    return f"Your file is now available at /get-dataframe/{new_file_id}"

@app.get("/get-dataframe/{new_file_id}")
async def get_dataframe(new_file_id: int):
    # Get data out of the DB and return it
    pass
```

Now, HTTP codes usually have an easier time dealing with single files at a time, and `201 Created` can only point towards a single new file, so the example above only deals with uploading one file at a time, but you get the idea.
